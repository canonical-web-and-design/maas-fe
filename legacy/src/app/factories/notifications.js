/* Copyright 2017 Canonical Ltd.  This software is licensed under the
 * GNU Affero General Public License version 3 (see the file LICENSE).
 *
 * MAAS Notifications Manager.
 *
 * Manages notifications in the browser. Uses RegionConnection to load
 * notifications, await new, updated, and deleted notifications, and to
 * dismiss them.
 */
import angular from "angular";

function NotificationsManager(RegionConnection, Manager) {
  function NotificationsManager() {
    Manager.call(this);

    this._pk = "id";
    this._handler = "notification";

    // Listen for notify events for the notification object.
    RegionConnection.registerNotifier(
      "notification",
      angular.bind(this, this.onNotify)
    );
  }

  NotificationsManager.prototype = new Manager();
  NotificationsManager.prototype.dismiss = function (notification) {
    return RegionConnection.callMethod("notification.dismiss", {
      id: notification.id,
    });
  };
  NotificationsManager.prototype.getItems = function () {
    return this._items.filter(({ ident }) => ident !== "release_notification");
  };

  return new NotificationsManager();
}

NotificationsManager.$inject = ["RegionConnection", "Manager"];

export default NotificationsManager;
