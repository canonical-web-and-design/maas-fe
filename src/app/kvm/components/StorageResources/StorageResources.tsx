import classNames from "classnames";

import StorageCards from "./StorageCards";
import StorageMeter from "./StorageMeter";

import type { KVMStoragePoolResources } from "app/kvm/types";
import type { Pod } from "app/store/pod/types";
import { formatBytes } from "app/utils";

type Props = {
  allocated: number; // B
  defaultPoolId?: Pod["default_storage_pool"];
  free: number; // B
  other?: number; // B
  pools: KVMStoragePoolResources;
};

const StorageResources = ({
  allocated,
  defaultPoolId,
  free,
  other = 0,
  pools,
}: Props): JSX.Element => {
  const freeStorage = formatBytes(free, "B");
  const totalStorage = formatBytes(allocated + other + free, "B");
  const singlePool = Object.keys(pools).length === 1;

  return (
    <div
      className={classNames("storage-resources", { "single-pool": singlePool })}
    >
      <div className="storage-resources__header">
        <h4 className="p-text--x-small-capitalised">Storage</h4>
        {!singlePool && (
          <div data-testid="storage-summary">
            <div className="u-nudge-left">
              <div className="p-text--x-small-capitalised u-text--muted u-no-margin--bottom">
                Total
              </div>
              <div className="u-align--right u-nudge-right u-sv1">
                {totalStorage.value}
                {totalStorage.unit}
              </div>
              <hr />
            </div>
            <div className="u-nudge-left">
              <div className="p-text--x-small-capitalised u-text--muted u-no-margin--bottom">
                Free
              </div>
              <div className="u-align--right u-nudge-right">
                {freeStorage.value}
                {freeStorage.unit}
              </div>
            </div>
          </div>
        )}
      </div>
      <div className="storage-resources__content">
        {singlePool ? (
          <StorageMeter defaultPoolId={defaultPoolId} pools={pools} />
        ) : (
          <StorageCards defaultPoolId={defaultPoolId} pools={pools} />
        )}
      </div>
    </div>
  );
};

export default StorageResources;
