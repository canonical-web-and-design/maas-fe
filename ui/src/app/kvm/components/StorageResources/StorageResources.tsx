import StorageCards from "./StorageCards";

import type { PodStoragePool } from "app/store/pod/types";
import { formatBytes } from "app/utils";

type Props = {
  storage: {
    allocated: number;
    free: number;
    pools: PodStoragePool[];
  };
};

const StorageResources = ({ storage }: Props): JSX.Element | null => {
  const freeStorage = formatBytes(storage.free, "B");
  const totalStorage = formatBytes(storage.allocated + storage.free, "B");

  return (
    <div className="storage-resources">
      <h4 className="storage-resources__header p-heading--small u-sv1">
        Storage
      </h4>
      <div className="storage-resources__content">
        <div className="storage-resources__usage">
          <div className="u-nudge-left">
            <div className="p-heading--small u-text--muted">Total</div>
            <div className="u-nudge-right u-sv1">
              {totalStorage.value}
              {totalStorage.unit}
            </div>
          </div>
          <div className="u-nudge-left">
            <div className="p-heading--small u-text--muted">Free</div>
            <div className="u-nudge-right u-sv1">
              {freeStorage.value}
              {freeStorage.unit}
            </div>
          </div>
        </div>
        <div className="storage-resources__pools">
          <StorageCards pools={storage.pools} />
        </div>
      </div>
    </div>
  );
};

export default StorageResources;
