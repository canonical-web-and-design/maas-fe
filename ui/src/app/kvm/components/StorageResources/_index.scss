@mixin StorageResources {
  .storage-resources {
    display: flex;
    flex-direction: row;
    padding: $spv--medium $sph--large;
  }

  .single-pool {
    flex-direction: column;
  }

  .storage-resources__content {
    display: flex;
    flex-direction: column;
    flex: 1;
  }

  .storage-resources__header {
    flex: 0;
    margin-right: 0;
  }

  @media only screen and (min-width: $breakpoint-small) {
    .storage-resources {
      flex-direction: row;
    }

    .storage-resources__header {
      flex: 1;
      margin-right: $sph--large;
    }

    .storage-resources__content {
      flex: 3;
    }
  }

  @media only screen and (min-width: $breakpoint-kvm-resources-card) {
    .storage-resources.single-pool {
      flex-direction: column;
    }

    .storage-resources__header {
      flex: 0;
      margin-right: 0;
    }

    .storage-resources__content {
      flex: 1;
    }
  }
}
