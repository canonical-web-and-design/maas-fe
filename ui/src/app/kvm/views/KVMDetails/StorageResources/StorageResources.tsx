import { useSelector } from "react-redux";

import StorageCards from "./StorageCards";
import StorageMeters from "./StorageMeters";

import podSelectors from "app/store/pod/selectors";
import type { Pod } from "app/store/pod/types";
import type { RootState } from "app/store/root/types";
import { formatBytes } from "app/utils";

type Props = { id: Pod["id"] };

const StorageResources = ({ id }: Props): JSX.Element | null => {
  const pod = useSelector((state: RootState) =>
    podSelectors.getById(state, id)
  );
  const sortedPools = useSelector((state: RootState) =>
    podSelectors.getSortedPools(state, Number(id))
  );
  const showCards = sortedPools.length >= 3;

  if (!pod) {
    return null;
  }

  const { allocated_other, allocated_tracked, free } = pod.resources.storage;
  const freeStorage = formatBytes(free, "B");
  const totalStorage = formatBytes(
    allocated_other + allocated_tracked + free,
    "B"
  );

  return (
    <div className="storage-resources">
      <h4 className="storage-resources__header p-heading--small u-sv1">
        Storage
      </h4>
      <div className="storage-resources__content">
        <div className="storage-resources__usage">
          <div className="u-nudge-left">
            <div className="p-heading--small u-text--muted">Total</div>
            <div className="u-nudge-right u-sv1">
              {totalStorage.value}
              {totalStorage.unit}
            </div>
          </div>
          <div className="u-nudge-left">
            <div className="p-heading--small u-text--muted">Free</div>
            <div className="u-nudge-right u-sv1">
              {freeStorage.value}
              {freeStorage.unit}
            </div>
          </div>
        </div>
        <div className="storage-resources__pools">
          {showCards ? (
            <StorageCards pools={sortedPools} />
          ) : (
            <StorageMeters pools={sortedPools} />
          )}
        </div>
      </div>
    </div>
  );
};

export default StorageResources;
