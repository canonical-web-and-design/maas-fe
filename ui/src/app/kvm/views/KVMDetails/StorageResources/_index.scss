@mixin StorageResources {
  .storage-resources {
    display: grid;
    grid-template-areas:
      "stats"
      "pools";
    grid-template-columns: 100%;
    grid-template-rows: min-content;
  }

  .storage-resources__stats {
    grid-area: stats;
    padding: $spv-inner--medium $sph-inner;
  }

  .storage-resources__usage {
    border-bottom: $border;
    display: flex;
    flex-direction: row;
  }

  .storage-resources__pools-container {
    grid-area: pools;
    overflow: auto;
    padding: 0 $sph-inner $spv-inner--medium $sph-inner;
  }

  .storage-resources__pool {
    display: flex;
    flex-direction: column;
    padding-top: $spv-inner--x-small;

    &:not(:last-of-type) {
      border-bottom: $border;
      padding-bottom: $spv-inner--x-small;
    }
  }

  .storage-resources__pool-name {
    flex: 1 0;
    margin-right: $sph-inner--large;
  }

  .storage-resources__pool-meter {
    flex: 3;
  }

  @media only screen and (min-width: $breakpoint-medium) {
    .storage-resources__pool {
      flex-direction: row;
    }
  }

  @media only screen and (min-width: $breakpoint-large) {
    .storage-resources {
      grid-template-areas: "stats pools";
      grid-template-columns: min-content 3fr;
      grid-template-rows: min-content;
    }

    .storage-resources__stats {
      padding: $spv-inner--medium 0 $spv-inner--medium $sph-inner;
    }

    .storage-resources__usage {
      border-bottom: 0;
      border-right: $border;
      flex-direction: column;
    }

    .storage-resources__pools-container {
      height: $sp-unit * 21;
      overflow: auto;
    }

    .storage-resources__pool {
      position: relative;
      top: $sp-unit * 3;
    }
  }
}
