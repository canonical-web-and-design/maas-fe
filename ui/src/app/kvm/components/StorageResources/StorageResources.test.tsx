import { shallow } from "enzyme";

import StorageResources from "./StorageResources";

import { podStoragePool as storagePoolFactory } from "testing/factories";

describe("StorageResources", () => {
  it("renders", () => {
    const storagePools = [
      storagePoolFactory({
        id: "0",
        name: "pool-0",
        path: "/path/0",
        total: 2,
        used: 1,
      }),
      storagePoolFactory({
        id: "1",
        name: "pool-1",
        path: "/path/1",
        total: 4,
        used: 3,
      }),
    ];
    const wrapper = shallow(
      <StorageResources
        storage={{ allocated: 5, free: 6, pools: storagePools }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
