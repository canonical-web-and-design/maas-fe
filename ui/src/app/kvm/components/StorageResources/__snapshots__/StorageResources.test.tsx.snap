// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StorageResources renders 1`] = `
<div
  className="storage-resources"
>
  <h4
    className="storage-resources__header p-heading--small u-sv1"
  >
    Storage
  </h4>
  <div
    className="storage-resources__content"
  >
    <div
      className="storage-resources__usage"
    >
      <div
        className="u-nudge-left"
      >
        <div
          className="p-heading--small u-text--muted"
        >
          Total
        </div>
        <div
          className="u-nudge-right u-sv1"
        >
          11
          B
        </div>
      </div>
      <div
        className="u-nudge-left"
      >
        <div
          className="p-heading--small u-text--muted"
        >
          Free
        </div>
        <div
          className="u-nudge-right u-sv1"
        >
          6
          B
        </div>
      </div>
    </div>
    <div
      className="storage-resources__pools"
    >
      <StorageCards
        pools={
          Array [
            Object {
              "available": 700000000000,
              "id": "0",
              "name": "pool-0",
              "path": "/path/0",
              "total": 2,
              "type": "lvm",
              "used": 1,
            },
            Object {
              "available": 700000000000,
              "id": "1",
              "name": "pool-1",
              "path": "/path/1",
              "total": 4,
              "type": "lvm",
              "used": 3,
            },
          ]
        }
      />
    </div>
  </div>
</div>
`;
